### DISCUSSION TITLE
Lesson 10: Deploy ERC20 token Error: Failed to get EIP-1559 fees

### DISCUSSION THREAD

**Discussion by LanceAddison** [Cyfrin/foundry-full-course-cu#1743] [Q&A] [answered]
Hey, I was trying to deploy my OurToken contract to the Sepolia network but I keep getting this error.

This is the command I am running:
```
forge script script/DeployOurToken.s.sol:DeployOurToken --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATE_KEY --broadcast --verify --etherscan-api-key $ETHERSCAN_API_KEY -vvvv
```

Here is the output with the error:
```
[⠒] Compiling...
No files changed, compilation skipped
Traces:
  [576781] DeployOurToken::run()
    ├─ [0] VM::envUint("PRIVATE_KEY") [staticcall]
    │   └─ ← [Return] <env var value>
    ├─ [0] VM::startBroadcast(<pk>)
    │   └─ ← [Return] 
    ├─ [518009] → new OurToken@0x6A7710b74ffC816Ac0E46D8fE3698EBEB4590ccE
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x3da99E808a621112b91d108De426439813cF6005, value: 1000000000000000000000000 [1e24])
    │   └─ ← [Return] 2130 bytes of code
    ├─ [0] VM::stopBroadcast()
    │   └─ ← [Return] 
    └─ ← [Return] OurToken: [0x6A7710b74ffC816Ac0E46D8fE3698EBEB4590ccE]


Script ran successfully.

== Return ==
0: contract OurToken 0x6A7710b74ffC816Ac0E46D8fE3698EBEB4590ccE

## Setting up 1 EVM.
==========================
Simulated On-chain Traces:

  [518009] → new OurToken@0x6A7710b74ffC816Ac0E46D8fE3698EBEB4590ccE
    ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x3da99E808a621112b91d108De426439813cF6005, value: 1000000000000000000000000 [1e24])
    └─ ← [Return] 2130 bytes of code

Error: 
Failed to get EIP-1559 fees
make: *** [deploy] Error 1
```

OurToken.sol contract:
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract OurToken is ERC20 {
    constructor(uint256 initialSupply) ERC20("OurToken", "OT") {
        _mint(msg.sender, initialSupply);
    }

    function burn(uint256 amount) public {
        _burn(msg.sender, amount);
    }
}
```

Code to my deploy script DeployOurToken.s.sol: 
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "../src/OurToken.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;
    uint256 public deployerKey;
    
    function run() external returns (OurToken) {
        if (block.chainid == 31337) {
            deployerKey = vm.envUint("DEFAULT_ANVIL_KEY");
        } else {
            deployerKey = vm.envUint("PRIVATE_KEY");
        }

        vm.startBroadcast(deployerKey);
        OurToken ot = new OurToken(INITIAL_SUPPLY);
        vm.stopBroadcast();
        return ot;
    }
}
```

When I run forge test with --fork-url $SEPOLIA_RPC_URL all the tests pass which really confused me as to why I am getting this error. I'm assuming it doesn't have to do with the contract or script its self. 

If anyone could help me figure this out I would really appreciate it.

**Comment by heeween** [comment, marked as answer]
I add --legacy and it works.

**Reply by LanceAddison** [reply]
Ok, I thought I ran that earlier (saw it somewhere online) and it didn't work but after running it now it worked. It was probably just something I messed up when trying to run the command. Thanks for your help!

Also, I'm not sure if you know but why exactly does adding --legacy make the error go away and why was the error even referring to eip-1559 when I thought this was supposed to be an erc-20 token? If you don't know it's all good. I'm just curious

**Reply by heeween** [reply]
I don't know either. The materials we can search from internet about foundry is not not too much. I thought it is an problem about foundry. I guess that.

**Reply by Jiutong666** [reply]
That's really helpful.

**Reply by musprodev** [reply]
> Ok, I thought I ran that earlier (saw it somewhere online) and it didn't work but after running it now it worked. It was probably just something I messed up when trying to run the command. Thanks for your help!
> 
> Also, I'm not sure if you know but why exactly does adding --legacy make the error go away and why was the error even referring to eip-1559 when I thought this was supposed to be an erc-20 token? If you don't know it's all good. I'm just curious

## Here is an answer @LanceAddison @heeween 

Before the introduction of *EIP-1559*, Ethereum transactions only had a single gas price field (`gasPrice`). so when you send a transaction, you specify this gas price, and miners include your transaction. 
> EIP-1559 transactions have a more complex fee structure that includes a base fee, a priority fee (tip), and a maximum fee (`maxFeePerGas` and `maxPriorityFeePerGas`).

So the reason it works is, not all ethereum rpcs or nodes might fully support EIP-1559 especially testnet nodes so by using the `--legacy` flag, you bypass the need for the node to provide *EIP-1559* fee data. And legacy transactions have a simpler fee model, requiring only the `gasPrice` parameter, which is easier for nodes to handle and less likely to encounter data retrieval issues.

**Reply by heeween** [reply]
thanks a lot. You explain very clearly.

**Reply by LanceAddison** [reply]
Thanks for the info. That makes it a lot clearer now

**Reply by 0xMaxii** [reply]
> > Ok, I thought I ran that earlier (saw it somewhere online) and it didn't work but after running it now it worked. It was probably just something I messed up when trying to run the command. Thanks for your help!
> > Also, I'm not sure if you know but why exactly does adding --legacy make the error go away and why was the error even referring to eip-1559 when I thought this was supposed to be an erc-20 token? If you don't know it's all good. I'm just curious
> 
> ## Here is an answer @LanceAddison @heeween
> Before the introduction of _EIP-1559_, Ethereum transactions only had a single gas price field (`gasPrice`). so when you send a transaction, you specify this gas price, and miners include your transaction.
> 
> > EIP-1559 transactions have a more complex fee structure that includes a base fee, a priority fee (tip), and a maximum fee (`maxFeePerGas` and `maxPriorityFeePerGas`).
> 
> So the reason it works is, not all ethereum rpcs or nodes might fully support EIP-1559 especially testnet nodes so by using the `--legacy` flag, you bypass the need for the node to provide _EIP-1559_ fee data. And legacy transactions have a simpler fee model, requiring only the `gasPrice` parameter, which is easier for nodes to handle and less likely to encounter data retrieval issues.

Nice one there muspro

**Reply by LanceAddison** [reply]
Thanks for the explanation.

**Reply by danibosna** [reply]
Thanks.

**Comment by EngrPips** [comment]
Hello @LanceAddison, `eip-1559` is the new standard that burns some part of the gas fee and then requires a `base fee` to have your transaction included in the block. You can read more about `eip-1559` here: https://ethereum.github.io/abm1559/notebooks/eip1559.html.

**Reply by LanceAddison** [reply]
Ah ok, thanks for the info. I will check it out

**Reply by EngrPips** [reply]
You are welcome.

**Reply by SamuelOsewa** [reply]
I tried the legacy flag and it works but when i search the contract address, etherscan doesn't show the address as a contract

**Reply by heeween** [reply]
> I tried the legacy flag and it works but when i search the contract address, etherscan doesn't show the address as a contract

Are you using --fork-url, but not --rpc-url? I think it is this param make the contract can't be shown in Etherscan.

**Reply by SamuelOsewa** [reply]
This is the command I used 
```
forge script --legacy --chain sepolia script/DeploySimpleStorage.s.sol --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATE_KEY --broadcast
```

**Reply by SamuelOsewa** [reply]
I've fixed it. I just had to add the :<script name> after the script path

**Reply by EngrPips** [reply]
Awesome, Glad you fixed it.

**Reply by jrmaktub** [reply]
> I've fixed it. I just had to add the :<script name> after the script path

Didn't you already include the script name above?

**Reply by EngrPips** [reply]
Maybe they didn't.
